src/main/java/org/mariadb/jdbc/internal/failover/impl/MastersSlavesListener.java

```diff
@@ -434,80 +434,88 @@ public void lockAndSwitchSecondary(Protocol newSecondaryProtocol) throws QueryEx
      */
     @Override
     public void switchReadOnlyConnection(Boolean mustBeReadOnly) throws QueryException {
-        if (mustBeReadOnly != currentReadOnlyAsked.get() && currentProtocol.inTransaction()) {
+        if (mustBeReadOnly != currentReadOnlyAsked && currentProtocol.inTransaction()) {
             throw new QueryException("Trying to set to read-only mode during a transaction");
         }
         checkWaitingConnection();
-        if (currentReadOnlyAsked.compareAndSet(!mustBeReadOnly, mustBeReadOnly)) {
```

Comment:
Oh, if your suggesting just keep the AtomicBoolean...yea, My personal option (style), is don't use AtomicX unless you need compare and swap operations, or unless you can take advantages of Atomic's weak set's.  Otherwise it's not worth the (granted very minor) memory usage.  It also may cause confusion by making it seem like a CAS operation may be safe to do on that variable, when instead it must be guarded by the lock.

So more style, than necessity if that is what you were trying to drive at.